subsys/net/ip/ipv6.c

```diff
@@ -1307,14 +1363,62 @@ static enum net_verdict handle_ns_input(struct net_pkt *pkt)
 		nd_opt_hdr = net_icmpv6_get_nd_opt_hdr(pkt, &ndopthdr);
 	}
 
-	ifaddr = net_if_ipv6_addr_lookup_by_iface(net_pkt_iface(pkt),
-						  &ns_hdr->tgt);
+	if (IS_ENABLED(CONFIG_NET_ROUTING)) {
+		ifaddr = net_if_ipv6_addr_lookup(&ns_hdr->tgt, NULL);
+	} else {
+		ifaddr = net_if_ipv6_addr_lookup_by_iface(net_pkt_iface(pkt),
+							  &ns_hdr->tgt);
+	}
+
 	if (!ifaddr) {
+		if (IS_ENABLED(CONFIG_NET_ROUTING)) {
```

Comment:
make all the below code block into an inline function. It becomes hairy with all these if/else imbrication and indentations starts to be messy due to 80 chars limit.

Reply:
This is not that bad actually indention wise. The problem with making this inline is that there are couple goto's here which make it actually cumbersome if this is inlined.